### PR TITLE
Feature flags in toolbar frontend

### PR DESCRIPTION
**Pull request by rcmarron**
## Changes

The front-end work to allow users to override their feature flags from the toolbar.

Dependent on #5824

https://user-images.githubusercontent.com/4813045/132799327-2ecce09d-e468-40b2-ab33-ee99fffff6fb.mov

Remaining items todo:
- [ ] Tests
- [ ] Get Posthog feature flags flowing into the toolbar to turn this feature on and off (super meta 🤔)

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests
- [ ] Migrations are safe to run at scale (e.g. PostHog Cloud) – present proof if not obvious
- [ ] New/changed UI is decent on smartphones (viewport width around 360px)

### CODE REVIEW

File: frontend/src/toolbar/flags/FeatureFlags.tsx [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+            </h1>
+            <Collapse>
+                {userFlags.map(({ feature_flag, value_for_user_without_override, override }) => {
+                    const hasVariants = (feature_flag.filters?.multivariate?.variants?.length || 0) > 0
+                    const flagEnabled = override ? !!override?.override_value : !!value_for_user_without_override
+                    const selectedVariant = hasVariants
+                        ? override
+                            ? override?.override_value
+                            : value_for_user_without_override
+                        : undefined
```
Comment (mariusandra) [inline frontend/src/toolbar/flags/FeatureFlags.tsx:24]: I think this would be a bit cleaner as a selector in a logic, though also fine to leave here.

File: frontend/src/toolbar/flags/featureFlagsLogic.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+            },
+        ],
+    }),
+
+    events: ({ actions }) => ({
+        afterMount: () => {
+            actions.getUserFlags()
+            ;(window['posthog'] as PostHog).onFeatureFlags((_, variants) => {
+                toolbarLogic.actions.updateFeatureFlags(variants)
+            })
```
Comment (mariusandra) [inline frontend/src/toolbar/flags/featureFlagsLogic.ts:74]: So this works when you're running the toolbar on the posthog app itself, but as soon as you're running the toolbar on someone else's website, we'd override the toolbar's feature flags (flags such as "are feature flags enabled in the toolbar?") with the site's feature flags ("show hero banner 2 on kea.js.org"). Fine for development for quick feedback, but should eventually be removed.

**Review by mariusandra — COMMENTED** [review]
Left some flyby comments, some to your, and some to my own code :).

I guess the missing bit is to get the feature flags into the toolbar, via the `editorParams`. The problem is these params need to be given in two places: `user.py` and `decide.py`, as 1) if you just open your site, we call `/decide` and show the toolbar... 2) if you click the toolbar in the sidebar in the posthog app, we do some URL redirection flow that gives the params in the URL after a `#hash` (also known as authorizing the toolbar).

Scope creep, but an improvement would be to create another of `/toolbar/boostrap` API endpoint, and make the toolbar get its real payload (e.g. the user's name, e-mail, ID, etc) from it. Some data (e.g. the token) should still be passed in the URL though..

**Comment by mariusandra** [comment]
Regarding the interface... I think this looks good, though I think it's visually useful to directly see the state of the flags as they are. I'm not sure what's the best way to communicate that. Maybe @clarkus has some quick suggestions?

**Comment by clarkus** [comment]
This looks like a great start! The radio button lists / option lists look solid, but the switch component at the top of the list would benefit from some clear labeling like "enable feature flag" or "disable feature flag". The placement of the "reset override" makes it feel related to that switch... I think hierarchically it's in the right spot, but it just needs a bit more separation. @rcmarron would you like me to draw out some quick changes for you or is that enough direction for now? Happy to make some time to support you on this.

**Comment by rcmarron** [comment]
Thanks @clarkus! I think I can run with that for now.

**Comment by rcmarron** [comment]
> Regarding the interface... I think this looks good, though I think it's visually useful to directly see the state of the flags as they are. I'm not sure what's the best way to communicate that. Maybe @clarkus has some quick suggestions?

I definitely agree. I had a version that showed the current value in the header, but it started to get pretty cluttered, and I was fighting the AntD component. I think I could easily put the current value in the expanded region below, and it could pair nicely with the 'reset override' button. I'll take a stab at that.

**Comment by clarkus** [comment]
Oh btw, feel free to ping me for a review on this when you're ready.

**Comment by rcmarron** [comment]
> Left some flyby comments, some to your, and some to my own code :).
> 
> I guess the missing bit is to get the feature flags into the toolbar, via the `editorParams`. The problem is these params need to be given in two places: `user.py` and `decide.py`, as 1) if you just open your site, we call `/decide` and show the toolbar... 2) if you click the toolbar in the sidebar in the posthog app, we do some URL redirection flow that gives the params in the URL after a `#hash` (also known as authorizing the toolbar).
> 
> Scope creep, but an improvement would be to create another of `/toolbar/boostrap` API endpoint, and make the toolbar get its real payload (e.g. the user's name, e-mail, ID, etc) from it. Some data (e.g. the token) should still be passed in the URL though..

@mariusandra I put some more thought into the "meta feature flag issue", and I have a couple of ideas that I wanted your thoughts on:

1) I might be thinking about it wrong, but I don't believe adding the feature flags in `user.py` and `decide.py` gets us what we want. If we add feature flags to those endpoints, they're going to be the customer's feature flags. (unless we do some messy piping of Posthog's flags in, but I don't think that works for self hosted customers). I think the same is true for a `/toolbar/bootstrap` endpoint. Definitely let me know if I'm thinking about this wrong.

2) It feels like another approach is to either add support for feature flags to `posthog-js-lite` or just have the toolbar hit Posthog's `/decide` endpoint. I feel like this depends on the future of `posthog-js-lite`. Do you have any thoughts between these 2?

3) Also, depending on our rollout strategy, we might be able to punt on this entire problem. If we're planning to release to internal and then roll out to 100%, I think we could have the tool using Posthog-js to get the feature flags (as it does now), and then when we want to roll out to 100%, remove the feature flag in code.

**Comment by mariusandra** [comment]
Hey, great analysis and you're absolutely right. The toolbar would indeed get the customer's flags, not ours. 

To get ours, we could fetch them out of the [`posthoganalytics`](https://posthog.com/docs/integrate/server/python#alternative-name) python package and pass to the toolbar, but that feels like a lot of extra steps to get the data to the toolbar.

We could indeed call `/decide` in posthog-js-lite, and eventually might do that, but that's quite out of scope here. We'd also need to add sending all the active flags with `capture`, etc. 

Hence, punting sounds fair, and actually not that bad of a compromise.

Basically, if any user anywhere has the `posthog-toolbar-feature-flags` feature flag enabled, we will show the feature flags in the toolbar. We'll have it enabled for us, and if anyone else wants to, they can also join the beta themselves.

When we actually prioritise working on the toolbar (e.g. new design/flow), we'll revisit the issue.

**Comment by mariusandra** [comment]
I think we can still improve the UI considerably

This is how it looks like for me:


My immediate wishlist:
- can see a dozen or more flags at a time
- see if those are enabled by just glancing at the list, including the selected variants
- change turn them on/off quickly
- see which flags I've modified and which I haven't

I do think the old version with switches (flag on/off) plus select boxes wasn't too bad:



(I don't have a gif ready with the variant selectboxes, but imagine them under the flag labels...)

To show if a flag has been modified or not, we could just give it a bolder background, add one line with a "overridden" tag, make the left border dark yellow, etc.

**Comment by clarkus** [comment]
@rcmarron Here's a quick idea for structuring that interface in a way that aligns with @mariusandra's wishlist

> - can see a dozen or more flags at a time
> - see if those are enabled by just glancing at the list, including the selected variants
> - change turn them on/off quickly
> - see which flags I've modified and which I haven't

Screen Shot 2021-09-13 at 11 21 28 AM

https://www.figma.com/file/9yWtngNb1AIuf6KmXaEPJA/App-doodles?node-id=1218%3A137379

This version includes a header with a dismiss action as well as an `active / total` count. This would show the number of enabled feature flags out of the available feature flags. Each item in the list can be toggled on /  off immediately via a switch component. For items that have sub-options, they can be expanded to show those options and select a new value. Items without sub-options have no expansion control. When an item is changed from its default value, we append a `custom` identifier to highlight this. Thoughts?

**Comment by samwinslow** [comment]
> see if those are enabled by just glancing at the list, including the selected variants

@clarkus — flyby comment: not sure this looks possible with the current design; what does `demo-banner-color` look like when the list of variants is collapsed, i.e. by default?

**Comment by samwinslow** [comment]
@mariusandra What do you think about making this a paid feature? (Or, more generally, making advanced feature flags a paid feature set?)

Seems like the power users of feature flags would likely be medium to large enterprises, like us, who have consistent CD pipelines and large enough user bases to the point where it makes sense to use complex flag logic. Happy to create a separate issue for this — it'll influence the way Growth talks about feature flags quite a bit.

**Comment by rcmarron** [comment]
@clarkus Thanks! This looks sweet!

One idea, that addresses @samwinslow 's comment is to show the actual value of the flag in the tag next to the flags name, and we could use a color bar like @mariusandra suggested to indicate the flag has been edited. I worry that it starts to be too much information in a small area, and as a result, is hard to parse. But it would be something like this:
image

**Comment by clarkus** [comment]
> @clarkus — flyby comment: not sure this looks possible with the current design; what does demo-banner-color look like when the list of variants is collapsed, i.e. by default?

It would look like the other collapsed items, just wouldn't have that `custom` identifier.

**Comment by clarkus** [comment]
> One idea, that addresses @samwinslow 's comment is to show the actual value of the flag in the tag next to the flags name, and we could use a color bar like @mariusandra suggested to indicate the flag has been edited. I worry that it starts to be too much information in a small area, and as a result, is hard to parse. But it would be something like this:

This could work, but it would depend on the character lengths of the flag names and the corresponding values. I also think that if we are going to color code, it'd be best to apply that to the entire item that is overridden. Are there character limits for either feature flag names or the variety of values they can contain?

Edit - here's an update that takes the inline-value idea and tries to make it scale a bit more. https://www.figma.com/file/9yWtngNb1AIuf6KmXaEPJA/App-doodles?node-id=1218%3A137818

Screen Shot 2021-09-13 at 1 45 24 PM

**Comment by rcmarron** [comment]
> > One idea, that addresses @samwinslow 's comment is to show the actual value of the flag in the tag next to the flags name, and we could use a color bar like @mariusandra suggested to indicate the flag has been edited. I worry that it starts to be too much information in a small area, and as a result, is hard to parse. But it would be something like this:
> 
> This could work, but it would depend on the character lengths of the flag names and the corresponding values. I also think that if we are going to color code, it'd be best to apply that to the entire item that is overridden. Are there character limits for either feature flag names or the variety of values they can contain?

Agreed on coloring the entire row.

Right now, the names have a max character length of 400, so essentially no limit for our purposes. And the values are not limited.

One other note is that the value only needs to be shown for 'enabled multi-variate feature flags'. If we limit it to those cases, it would remove some of the clutter.
image

**Comment by rcmarron** [comment]
> Edit - here's an update that takes the inline-value idea and tries to make it scale a bit more. https://www.figma.com/file/9yWtngNb1AIuf6KmXaEPJA/App-doodles?node-id=1218%3A137818

This looks great! I'll run with it.

**Comment by clarkus** [comment]
One more option that breaks labels and values into distinct lines. This would make each option taller, but would give each value the full width of the component to display the value. 

Screen Shot 2021-09-13 at 1 55 16 PM

**Comment by rcmarron** [comment]
> One more option that breaks labels and values into distinct lines. This would make each option taller, but would give each value the full width of the component to display the value.

I like this. 2 Thoughts:
* The value of the flag is probably less interesting than the flag name itself. Most often, I'm probably scanning this list looking for my flag's name. I wonder if we should flip the bold
* Right now, multi-variate flags are pretty uncommon (they're behind a feature flag), so most flags won't have the expand icon on the left. It feels like that icon gives a major "anchor" for the row, and I'm wondering if we need something there when the row isn't a mult-variate flag.

**Comment by clarkus** [comment]
After talking through the details of the issue with @rcmarron we landed on this as a starting point. Note that as feature flags are enabled, the option-list will expand to show all options. The assumption is that there will not be a lot of multi-variate flags, and when there are, there will not be a lot of options (less than 10). If usage contradicts this assumption, we might have to adjust this design back to a previous version that had more complexity, but allowed for the collapsing of overridden flags. 

https://www.figma.com/file/9yWtngNb1AIuf6KmXaEPJA/App-doodles?node-id=1218%3A137818

Screen Shot 2021-09-13 at 2 23 34 PM

**Comment by mariusandra** [comment]
Love the new look!

One thing to keep in mind is that the flags will be in various states of enabled/disabled, and the overrides can also set them into various states of enabled/disabled (within the boundary of the flag type). The mockups here all seem to align "overridden" with "enabled", though you might also override a flag off. 

I'm not sure if that changes anything though.

**Comment by rcmarron** [comment]
This is ready for review. A couple of notes:

* This feature lives behind the feature flag `posthog-toolbar-feature-flags`, and can only be rolled out internally right now. This is because toolbar does not have access to Posthog's feature flags. Rather, the toolbar has access to the customer's feature flags, so right now, this feature will only work internally (because in that case, the "customer" is Posthog and we have the flag `posthog-toolbar-feature-flags` set). There is a bit of work to give the toolbar access to feature flags, and when we want to roll this out, we can either do that work or just remove the feature flag that it's behind.
* The toolbar can definitely use a design pass beyond this feature. I tried to make the existing panels match the styles for the new feature flag panel, but it was a bit tricky. (see the pics below for where I landed) @clarkus I'd love your thoughts here.




* Here is a quick demo of the feature 

https://user-images.githubusercontent.com/4813045/133349678-ff59ed70-a703-4f1f-a659-5b7d46322575.mov

**Comment by clarkus** [comment]
@rcmarron This looks good to me. I am tracking longer term toolbar improvements at https://github.com/PostHog/posthog/issues/2370. If you had any specific items you think should be addressed, can you add them to that issue?

**Review by mariusandra — APPROVED** [review]
This looks amazing! 

I have just two comments regarding the nemesis of all macOS frontend developers: scrollbars 🤣 .

1. The actions view has a scrollbar:


I actually recommend everyone working in frontend enable this setting to always catch them:

Screenshot 2021-09-15 at 12 05 41

The problem is all Windows computers always show scrollbars, so it'll look really ugly for those users... and we must prevent that.

2. The scrollbar here is also slightly off. Could it be at the very right of the window?

image
